File: packages/template-python/src/deepagents_template/surfaces/cli/extract_content.py
"""Content extraction helper — extracts text content from agent responses."""

from __future__ import annotations

from typing import Any


def extract_content(response: Any) -> str:
    """Extract text content from an agent response."""
    if response is None:
        return ""
    if isinstance(response, str):
        return response
    if isinstance(response, dict):
        content = response.get("content", response.get("text", ""))
        if isinstance(content, str):
            return content
        if isinstance(content, list):
            parts = []
            for block in content:
                if isinstance(block, dict) and block.get("type") == "text":
                    parts.append(block.get("text", ""))
            return "\n".join(parts)
        return str(content)
    return str(response)

File: packages/template/src/surfaces/cli/extract-content.ts
/**
 * Extract text content from an agent invoke response.
 * Handles various response formats from different deepagents versions.
 */
export function extractContent(response: unknown): string {
  if (!response) return "(no response)";

  if (typeof response === "string") return response;

  // LangChain message array response
  if (Array.isArray(response)) {
    return response
      .map((m: unknown) => extractContent(m))
      .filter(Boolean)
      .join("\n");
  }

  // Object with messages array
  const r = response as { messages?: unknown[]; content?: unknown; text?: unknown };
  if (Array.isArray(r.messages)) {
    return r.messages
      .map((m: unknown) => extractContent(m))
      .filter(Boolean)
      .join("\n");
  }

  if (typeof r.content === "string") return r.content;
  if (typeof r.text === "string") return r.text;

  // Array of content blocks
  if (Array.isArray(r.content)) {
    return r.content
      .map((b: unknown) => {
        if (typeof b === "string") return b;
        const block = b as { type?: string; text?: string };
        if (block.type === "text" && typeof block.text === "string") return block.text;
        return "";
      })
      .filter(Boolean)
      .join("\n");
  }

  // Fallback: stringify
  try {
    return JSON.stringify(response, null, 2);
  } catch {
    return String(response);
  }
}
